flowapi/flowapi/permissions.py

```diff
@@ -0,0 +1,358 @@
+        Schema dict to turn into scopes list
+
+    Yields
+    ------
+    str
+        Scope strings
+
+    Examples
+    --------
+    >>>list(schema_to_scopes({"FlowmachineQuerySchema": {"oneOf": [{"$ref": "DUMMY"}]},"DUMMY": {"properties": {"query_kind": {"enum": ["dummy"]}}},},))
+    ["get_result:dummy", "run:dummy", "get_result:available_dates"],
+    """
+    return make_scopes(
+        queries=ResolvingParser(spec_string=dumps(schema)).specification["components"][
+            "schemas"
+        ]["FlowmachineQuerySchema"]
+    )
+
+
+def expand_scopes(*, scopes: List[str]) -> str:
+    """
+    Expand up a list of compact scopes to full scopes
+
+    Parameters
+    ----------
+    scopes : list of str
+        Compressed scopes to expand
+
+    Yields
+    ------
+    str
+        A scope string
+
+    """
+    for scope in scopes:
+        parts = scope.split(":")
+        ps = (x.split(",") for x in parts)
+        yield from (":".join(x) for x in product(*ps))
+
+
+def q_to_scope_atoms(
+    *, query: dict, argument_names_to_extract: List[str] = ["aggregation_unit"],
+) -> List[Union[str, List[str]]]:
+    """
+    Convert a query to scope 'atoms', each of which would be a contiguous
+    part of a scope string.
+
+    Parameters
+    ----------
+    query : dict
+        Query to atomise
+    argument_names_to_extract : list of str
+        Arguments to pull out
+
+    Returns
+    -------
+    list of str
+       Nested list of atoms
+    """
+    paths = []
+    if "query_kind" in query:
+        paths = paths + [query["query_kind"]]
+    for k, v in sorted(query.items()):
+        if k in argument_names_to_extract:
+            paths = paths + [k, v]
+        if isinstance(v, dict):
+            atoms = q_to_scope_atoms(query=v)
+            if len(atoms) > 0:
+                paths = paths + [k] + atoms
+        if isinstance(v, list):
+            child_paths = set()
+            for x in v:
+                if isinstance(x, dict):
+                    child_paths.add(tuple(q_to_scope_atoms(query=x)))
+            for child in child_paths:
+                if len(child) > 0:
+                    paths = paths + [k] + list(child)
+    return paths
+
+
+def q_to_subscopes(
+    *, query: dict, argument_names_to_extract: List[str] = ["aggregation_unit"],
+):
+    """
+    Translate a query into contiguous 'subscopes', which when combined in
+    the correct order form a complete scope.
+
+    Parameters
+    ----------
+    query : dict
+        Query to translate to subscopes
+     argument_names_to_extract : list of str
+        Arguments to pull out
+    Returns
+    -------
+    list of str
+        List of subscopes
+
+    """
+    atoms = q_to_scope_atoms(
+        query=query, argument_names_to_extract=argument_names_to_extract
+    )
+    top_level = [x for x in atoms if isinstance(x, str)]
+    return [":".join(top_level)] + [
+        ":".join(flatten(x)) for x in atoms if isinstance(x, list)
```

Comment:
I'm not sure about this. Separating the top_level atoms from those that require flattening will rearrange the order in the final scope string, which could lead to ambiguous results if two child queries within the query have parameters with the same name (as will certainly be the case with `aggregation_unit` in many cases, and quite possibly other parameter names as well).

Having thought about this for a bit, I've come up with the following proposal for a slightly different format for scope strings allowing for nested sub-scopes, which might be worth considering:

Basic form of a scope is `'action:query_kind'`, where 'action' is either 'run' or 'get_result'.

Sub-scopes for 'query_kind' are represented as `'action:query_kind{sub_scopes}'`, where 'sub_scopes' is a ':'-separated list of key:value pairs (e.g. `'action:query_kind{param1:value1:param2:value2}'`).

Each key in 'sub_scopes' is the name of a parameter of a 'query_kind' query, and each value is a value for that parameter, which could be an aggregation unit or a child query kind.

If a query kind has no child queries or aggregation unit, the scope reduces to `'action:query_kind{}'`, which is equivalent to `'action:query_kind'` (without the curly braces).

Child queries can also have sub-scopes. For example, assume 'query_kind_1' takes two child queries as parameters 'param_a' and 'param_b', and 'query_kind_2' and 'query_kind_3' each have an aggregation unit. The 'run' scope for a query of kind 'query_kind_1' with 'param_a' a query of kind 'query_kind_2' and 'param_b' a query of kind 'query_kind_3', both having aggregation unit 'agg', would be `'run:query_kind_1{param_a:query_kind_2{aggregation_unit:agg}:param_b:query_kind_3{aggregation_unit:agg}}'`.

Splitting a scope string into its components (e.g. for squashing/expanding) would be a bit less straightforward - can't just split on ':', we'd need to do some regex-ing - but I don't think this would be too much of a problem.

This should be more squashable than the current flat list, too. At the moment, two scopes
```
[
    'run:query_kind:aggregation_unit:admin1',
    'run:query_kind:aggregation_unit:admin2',
]
```
could be squashed to
```
['run:query_kind:aggregation_unit:admin1,admin2']
```
but
```
[
    'run:query_kind_1:<parameter:value pairs>',
    'run:query_kind_2:<parameter:value pairs>',
]
```
can't be squashed (unless the parameters for query_kind_1 and query_kind_2 happen to be identical). Using nested sub-scopes,
```
[
    'run:query_kind_1{<parameter:value pairs>}',
    'run:query_kind_2{<parameter:value pairs>}',
]
```
could be squashed to
```
['run:query_kind_1{<parameter:value pairs>},query_kind_2{<parameter:value pairs>}']
```
Admittedly the squashing in this example doesn't save us much, but in a more complex query (e.g. flows) the additional squashing could be significant. Maximal squashing/expanding would require an iterative or recursive approach, but again I can't see a problem with this.

If it's necessary to represent a scope as a python object, I'd suggest each value could be parsed to a tuple `(kind, sub_scopes)`, where `sub_scopes` is a dict (or frozenset of dict items, if it needs to be immutable), so parsing the example scope above would give
```python
"run",
(
    "query_kind_1",
    {
        "param_a": ("query_kind_2", {"aggregation_unit": ("agg", {})}),
        "param_b": ("query_kind_3", {"aggregation_unit": ("agg", {})}),
    },
)
```

Reply:
It may just be that squashing it into a string isn't tenable and we need to stick with the current approach of sending around dicts. I was wondering something similar on the scope-string 'grammar' front, but I feel like once one gets any further into that territory we might as well just go for (as we do now) encoding directly in JSON and then squashing it some other way - already getting some distance from human readable in what's in this PR.

Reply:
Hmm, perhaps. Personally I think the nested-sub-scopes form is a good middle-ground - it's not far off JSON, but it's more concise, and in my opinion no less readable than either a JSON or ':'-separated flat list representation. It's also squashable enough that the full set of 'run,get_result' scopes (i.e. everything except 'get_result:available_dates') could be expressed as a single string.

A single (i.e. not squashed) nested scope string could be reversibly converted to flat-list form using dot-delimited parameter names, e.g.
```
run:query_kind_1{param_a:query_kind_2{aggregation_unit:agg}:param_b:query_kind_3{aggregation_unit:agg}}
```
could be expressed as
```
run:query_kind_1:param_a:query_kind_2:param_a.aggregation_unit:agg:param_b:query_kind_3:param_b.aggregation_unit:agg
```
which still resolves the issue of ambiguity when different queries have the same parameter names, and could be more convenient for rendering in FlowAuth. I think I'd prefer the nested form for inclusion in the API spec and tokens, though, because it's more concise and more squashable (squidgier?).